073-decouple-ca-management-from-end-entity-certificate-management.md

```diff
@@ -0,0 +1,250 @@
+
+```
+  UNTRUSTED // not yet trusted everywhere it needs to be 
+   |
+   v
+  TRUSTED_UNUSED // trusted everywhere it needs to be, but no EE certificates yet issued with this CA certificate
+   |
+   v
+  TRUSTED_IN_USE // trusted everywhere, EE certificates issued using this CA certificate
+   |
+   v
+  PHASE_OUT // trusted everywhere, but new end-entity certificates will not be issued using this CA certificate
+   |
+   v
+  ZOMBIE // still trusted by some component, but with no issued end-entity certificates in current use.
+```
+
+When a new cluster is first created the flow is:
+
+1. Either user or CaReconciler creates the Cluster and Client CA certificates and keys and stores them in Kubernetes Secrets.
+2. The CaReconciler takes the Cluster CA certificate and places it in the new shared Kubernetes Secret with the state UNTRUSTED
+3. At the beginning of the reconcile loop of components, the cluster operator generates the Cluster Operator certificate and stores it in the <CLUSTER_NAME>-cluster-operator-certs Kubernetes Secret
+   It does this even though the state is UNTRUSTED, because the Kubernetes Secret was missing
+4. The component reconcilers (KafkaReconciler, ZooKeeperReconciler, EntityOperatorReconciler etc):
+   1. generate their certificates, store them in Kubernetes Secrets
+   2. create their pods annotated with the certificates Kubernetes Secret and new shared Kubernetes Secret mounted onto the pod
+5. At the end of the reconcile loop of components, the cluster operator updates the shared Kubernetes Secret to change the state of the CA certificate to TRUSTED_IN_USE
+
+When the CA private key is updated (either because it has expired or the user has incremented the generation to indicate they've manually updated the Kubernetes Secret):
+
+1. If Strimzi is managing the CA, the CaReconciler renews the Cluster and Client CAs certificates and keys. The old certificate and key can be removed from the Kubernetes Secret.
+2. The CaReconciler takes the Cluster CA certificate and places it in the new shared Kubernetes Secret with the state UNTRUSTED. It also updates the old CA certificate to have the state PHASE_OUT.
+3. The component reconcilers (KafkaReconciler, ZooKeeperReconciler, EntityOperatorReconciler etc) see the new UNTRUSTED CA certificate and:
+   1. Roll the pods to trust the new CA certificate
+4. At the end of the reconcile loop the cluster operator updates the shared Kubernetes Secret to change the state of the CA certificate to TRUSTED_UNUSED
+5. This change triggers a new reconcile loop
+6. At the beginning of the reconcile loop the cluster operator sees there is a CA certificate in the state TRUSTED_UNUSED and regenerates the Cluster Operator certificate, signed by the new CA key and stores it in the <CLUSTER_NAME>-cluster-operator-certs Kubernetes Secret
+7. The component reconcilers (KafkaReconciler, ZooKeeperReconciler, EntityOperatorReconciler etc) see the new TRUSTED_UNUSED CA certificate and:
```

Comment:
How does a component `see the new TRUSTED_UNUSED CA certificate and get the new server cert`? The motivation suggests that arbitrary trusted certs come with the private key without prior knowledge of the operator. So you need to first get the _future_ server certificate to figure out you need to trust its CA and only then once the other components trust it (that is n on-trivial for example also because of the TO->CC communication) you can start using it. So I think there is a huge impact on the flow but also on the used Secrets etc. (that said, as my comments on the motivation suggest, I do not think it should really work that way => but it does not seem to fit together.)